man/tabs.1
.\"***************************************************************************
.\" Copyright (c) 2008-2009,2010 Free Software Foundation, Inc.              *
.\"                                                                          *
.\" Permission is hereby granted, free of charge, to any person obtaining a  *
.\" copy of this software and associated documentation files (the            *
.\" "Software"), to deal in the Software without restriction, including      *
.\" without limitation the rights to use, copy, modify, merge, publish,      *
.\" distribute, distribute with modifications, sublicense, and/or sell       *
.\" copies of the Software, and to permit persons to whom the Software is    *
.\" furnished to do so, subject to the following conditions:                 *
.\"                                                                          *
.\" The above copyright notice and this permission notice shall be included  *
.\" in all copies or substantial portions of the Software.                   *
.\"                                                                          *
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS  *
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF               *
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.   *
.\" IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,   *
.\" DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR    *
.\" OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR    *
.\" THE USE OR OTHER DEALINGS IN THE SOFTWARE.                               *
.\"                                                                          *
.\" Except as contained in this notice, the name(s) of the above copyright   *
.\" holders shall not be used in advertising or otherwise to promote the     *
.\" sale, use or other dealings in this Software without prior written       *
.\" authorization.                                                           *
.\"***************************************************************************
.\"
.\" $Id: tabs.1,v 1.8 2010/12/04 18:40:45 tom Exp $
.TH @TABS@ 1 ""
.ds n 5
.SH NAME
\fBtabs\fR \- set tabs on a terminal
.SH SYNOPSIS
\fBtabs\fR [\fB\-v\fR[\fIn\fR]] [\fB\-ahuUV\fR] \fIfile...\fR
.br
.SH DESCRIPTION
.PP
The \fBtabs\fP program clears and sets tab-stops on the terminal.
This uses the terminfo \fIclear_all_tabs\fP and \fIset_tab\fP capabilities.
If either is absent, \fBtabs\fP is unable to clear/set tab-stops.
The terminal should be configured to use hard tabs, e.g.,
.sp
.RS
stty tab0
.RE
.SH OPTIONS
.SS General Options
.TP 5
.BI \-T "name"
Tell \fBtabs\fP which terminal type to use.
If this option is not given, \fBtabs\fP will use the \fB$TERM\fP
environment variable.
If that is not set, it will use the \fIansi+tabs\fP entry.
.TP 5
.B \-d
The debugging option shows a ruler line, followed by two data lines.
The first data line shows the expected tab-stops marked with asterisks.
The second data line shows the actual tab-stops, marked with asterisks.
.TP 5
.B \-n
This option tells \fBtabs\fP to check the options and run any debugging
option, but not to modify the terminal settings.
.PP
The \fBtabs\fP program processes a single list of tab stops.
The last option to be processed which defines a list is the one that
determines the list to be processed.
.SS Implicit Lists
Use a single number as an option, e.g., "\fB\-5\fP" to set tabs at the given
interval (in this case 1, 6, 11, 16, 21, etc.).  Tabs are repeated up to
the right margin of the screen.
.PP
Use "\fB\-0\fP" to clear all tabs.
.PP
Use "\fB\-8\fP" to set tabs to the standard interval.
.SS Explicit Lists
An explicit list can be defined after the options (this does not use a "\-").
The values in the list must be in increasing numeric order, and greater than
zero.  They are separated by a comma or a blank, for example,
.sp
.RS
tabs 1,6,11,16,21
.br
tabs 1 6 11 16 21
.RE
Use a '+' to treat a number as an increment relative to the previous value,
e.g.,
.sp
.RS
tabs 1,+5,+5,+5,+5
.RE
which is equivalent to the 1,6,11,16,21 example.
.SS Predefined Tab-Stops
X/Open defines several predefined lists of tab stops.
.TP 5
.B \-a
Assembler, IBM S/370, first format
.TP 5
.B \-a2
Assembler, IBM S/370, second format
.TP 5
.B \-c
COBOL, normal format
.TP 5
.B \-c2
COBOL compact format
.TP 5
.B \-c3
COBOL compact format extended
.TP 5
.B \-f
FORTRAN
.TP 5
.B \-p
PL/I
.TP 5
.B \-s
SNOBOL
.TP 5
.B \-u
UNIVAC 1100 Assembler
.SH PORTABILITY
.PP
X/Open describes a \fB+m\fP option, to set a terminal's left-margin.
Very few of the entries in the terminal database provide this capability.
.PP
The \fB\-d\fP (debug) and \fB\-n\fP (no-op) options are extensions not provided
by other implementations.
.PP
Documentation for other implementations states that there is a limit on the
number of tab stops.  While some terminals may not accept an arbitrary number
of tab stops, this implementation will attempt to set tab stops up to the
right margin of the screen, if the given list happens to be that long.
.SH SEE ALSO
\fB@TSET@\fR(1),
\fB@INFOCMP@\fR(1M),
\fBcurses\fR(3X),
\fBterminfo\fR(\*n).
.PP
This describes \fBncurses\fR
version @NCURSES_MAJOR@.@NCURSES_MINOR@ (patch @NCURSES_PATCH@).


NEWS
-------------------------------------------------------------------------------
-- Copyright (c) 1998-2010,2011 Free Software Foundation, Inc.               --
--                                                                           --
-- Permission is hereby granted, free of charge, to any person obtaining a   --
-- copy of this software and associated documentation files (the             --
-- "Software"), to deal in the Software without restriction, including       --
-- without limitation the rights to use, copy, modify, merge, publish,       --
-- distribute, distribute with modifications, sublicense, and/or sell copies --
-- of the Software, and to permit persons to whom the Software is furnished  --
-- to do so, subject to the following conditions:                            --
--                                                                           --
-- The above copyright notice and this permission notice shall be included   --
-- in all copies or substantial portions of the Software.                    --
--                                                                           --
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS   --
-- OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF                --
-- MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN --
-- NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,       --
-- DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR     --
-- OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE --
-- USE OR OTHER DEALINGS IN THE SOFTWARE.                                    --
--                                                                           --
-- Except as contained in this notice, the name(s) of the above copyright    --
-- holders shall not be used in advertising or otherwise to promote the      --
-- sale, use or other dealings in this Software without prior written        --
-- authorization.                                                            --
-------------------------------------------------------------------------------
-- $Id: NEWS,v 1.1682 2011/04/04 00:02:42 tom Exp $
-------------------------------------------------------------------------------

This is a log of changes that ncurses has gone through since Zeyd started
working with Pavel Curtis' original work, pcurses, in 1992.

Changes through 1.9.9e are recorded by Zeyd M Ben-Halim.
Changes since 1.9.9e are recorded by Thomas E Dickey.

Contributors include those who have provided patches (even small ones), as well
as those who provide useful information (bug reports, analyses).  Changes with
no cited author are the work of Thomas E Dickey (TD).

A few contributors are given in this file by their initials.
They each account for one percent or more of the changes since 1.9.9e.
See the AUTHORS file for the corresponding full names.

Changes through 1.9.9e did not credit all contributions;
it is not possible to add this information.

20110404 5.9 release for upload to ftp.gnu.org

20110402
	+ various build-fixes for the rpm/dpkg scripts.
	+ add "--enable-rpath-link" option to Ada95/configure, to allow
	  packages to suppress the rpath feature which is normally used for
	  the in-tree build of sample programs.
	+ corrected definition of libdir variable in Ada95/src/Makefile.in,
	  needed for rpm script.
	+ add "--with-shared" option to Ada95/configure script, to allow
	  making the C-language parts of the binding use appropriate compiler
	  options if building a shared library with gnat.

20110329
	> portability fixes for Ada95 binding:
	+ add configure check to ensure that SIGINT works with gnat.  This is
	  needed for the "rain" sample program.  If SIGINT does not work, omit
	  that sample program.
	+ correct typo in check of $PKG_CONFIG variable in Ada95/configure
	+ add ncurses_compat.c, to supply functions used in the Ada95 binding
	  which were added in 5.7 and later.
	+ modify sed expression in CF_NCURSES_ADDON to eliminate a dependency
	  upon GNU sed.

20110326
	+ add special check in Ada95/configure script for ncurses6 reentrant
	  code.
	+ regen Ada html documentation.
	+ build-fix for Ada shared libraries versus the varargs workaround.
	+ add rpm and dpkg scripts for Ada95 and test directories, for test
	  builds.
	+ update test/configure macros CF_CURSES_LIBS, CF_XOPEN_SOURCE and
	  CF_X_ATHENA_LIBS.
	+ add configure check to determine if gnat's project feature supports
	  libraries, i.e., collections of .ali files.
	+ make all dereferences in Ada95 samples explicit.
	+ fix typo in comment in lib_add_wch.c (patch by Petr Pavlu).
	+ add configure check for, ifdef's for math.h which is in a separate
	  package on Solaris and potentially not installed (report by Petr
	  Pavlu).
	> fixes for Ada95 binding (Nicolas Boulenguez):
	+ improve type-checking in Ada95 by eliminating a few warning-suppress
	  pragmas.
	+ suppress unreferenced warnings.
	+ make all dereferences in binding explicit.

20110319
	+ regen Ada html documentation.
	+ change order of -I options from ncurses*-config script when the
	  --disable-overwrite option was used, so that the subdirectory include
	  is listed first.
	+ modify the make-tar.sh scripts to add a MANIFEST and NEWS file.
	+ modify configure script to provide value for HTML_DIR in
	  Ada95/gen/Makefile.in, which depends on whether the Ada95 binding is
	  distributed separately (report by Nicolas Boulenguez).
	+ modify configure script to add -g and/or -O3 to ADAFLAGS if the
	  CFLAGS for the build has these options.
	+ amend change from 20070324, to not add 1 to the result of getmaxx
	  and getmaxy in the Ada binding (report by Nicolas Boulenguez for
	  thread in comp.lang.ada).
	+ build-fix Ada95/samples for gnat 4.5
	+ spelling fixes for Ada95/samples/explain.txt
	> fixes for Ada95 binding (Nicolas Boulenguez):
	+ add item in Trace_Attribute_Set corresponding to TRACE_ATTRS.

progs/tabs.c
/****************************************************************************
 * Copyright (c) 2008-2009,2010 Free Software Foundation, Inc.              *
 *                                                                          *
 * Permission is hereby granted, free of charge, to any person obtaining a  *
 * copy of this software and associated documentation files (the            *
 * "Software"), to deal in the Software without restriction, including      *
 * without limitation the rights to use, copy, modify, merge, publish,      *
 * distribute, distribute with modifications, sublicense, and/or sell       *
 * copies of the Software, and to permit persons to whom the Software is    *
 * furnished to do so, subject to the following conditions:                 *
 *                                                                          *
 * The above copyright notice and this permission notice shall be included  *
 * in all copies or substantial portions of the Software.                   *
 *                                                                          *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS  *
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF               *
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.   *
 * IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,   *
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR    *
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR    *
 * THE USE OR OTHER DEALINGS IN THE SOFTWARE.                               *
 *                                                                          *
 * Except as contained in this notice, the name(s) of the above copyright   *
 * holders shall not be used in advertising or otherwise to promote the     *
 * sale, use or other dealings in this Software without prior written       *
 * authorization.                                                           *
 ****************************************************************************/

/****************************************************************************
 *  Author: Thomas E. Dickey                        2008                    *
 ****************************************************************************/

/*
 * tabs.c --  set terminal hard-tabstops
 */

#define USE_LIBTINFO
#include <progs.priv.h>

MODULE_ID("$Id: tabs.c,v 1.19 2010/10/23 22:26:01 tom Exp $")

static void usage(void) GCC_NORETURN;

static int max_cols;

static int
putch(int c)
{
    return putchar(c);
}

static void
do_tabs(int *tab_list)
{
    int last = 1;
    int stop;

    putchar('\r');
    while ((stop = *tab_list++) > 0) {
	if (last < stop) {
	    while (last++ < stop) {
		if (last > max_cols)
		    break;
		putchar(' ');
	    }
	}
	if (stop <= max_cols) {
	    tputs(tparm(set_tab, stop), 1, putch);
	    last = stop;
	} else {
	    break;
	}
    }
    putchar('\n');
}

static int *
decode_tabs(const char *tab_list)
{
    int *result = typeCalloc(int, strlen(tab_list) + (unsigned) max_cols);
    int n = 0;
    int value = 0;
    int prior = 0;
    int ch;

    if (result != 0) {
	while ((ch = *tab_list++) != '\0') {
	    if (isdigit(UChar(ch))) {
		value *= 10;
		value += (ch - '0');
	    } else if (ch == ',') {
		result[n] = value + prior;
		if (n > 0 && result[n] <= result[n - 1]) {
		    fprintf(stderr,
			    "tab-stops are not in increasing order: %d %d\n",
			    value, result[n - 1]);
		    free(result);
		    result = 0;
		    break;
		}
		++n;
		value = 0;
		prior = 0;
	    } else if (ch == '+') {
		if (n)
		    prior = result[n - 1];
	    }
	}
    }

    if (result != 0) {
	/*
	 * If there is only one value, then it is an option such as "-8".
	 */
	if ((n == 0) && (value > 0)) {
	    int step = value;
	    while (n < max_cols - 1) {
		result[n++] = value;
		value += step;
	    }
	}

	/*
	 * Add the last value, if any.
	 */
	result[n++] = value + prior;
	result[n] = 0;
    }
    return result;
}

static void
print_ruler(int *tab_list)
{
    int last = 0;
    int stop;
    int n;

    /* first print a readable ruler */
    for (n = 0; n < max_cols; n += 10) {
	int ch = 1 + (n / 10);
	char buffer[20];
	sprintf(buffer, "----+----%c",
		((ch < 10)
		 ? (ch + '0')
		 : (ch + 'A' - 10)));
	printf("%.*s", ((max_cols - n) > 10) ? 10 : (max_cols - n), buffer);
    }
    putchar('\n');

    /* now, print '*' for each stop */
    for (n = 0, last = 0; (tab_list[n] > 0) && (last < max_cols); ++n) {
	stop = tab_list[n];
	while (++last < stop) {
	    if (last <= max_cols) {
		putchar('-');
	    } else {
		break;
	    }
	}
	if (last <= max_cols) {
	    putchar('*');
	    last = stop;
	} else {
	    break;
	}
    }
    while (++last <= max_cols)
	putchar('-');
    putchar('\n');
}

/*
 * Write an '*' on each tabstop, to demonstrate whether it lines up with the
 * ruler.
 */
static void
write_tabs(int *tab_list)
{
    int stop;

    while ((stop = *tab_list++) > 0 && stop <= max_cols) {
	fputs((stop == 1) ? "*" : "\t*", stdout);
    };
    /* also show a tab _past_ the stops */
    if (stop < max_cols)
	fputs("\t+", stdout);
    putchar('\n');
}

/*
 * Trim leading/trailing blanks, as well as blanks after a comma.
 * Convert embedded blanks to commas.
 */
static char *
trimmed_tab_list(const char *source)
{
    char *result = strdup(source);
    int ch, j, k, last;

    if (result != 0) {
	for (j = k = last = 0; result[j] != 0; ++j) {
	    ch = UChar(result[j]);
	    if (isspace(ch)) {
		if (last == '\0') {
		    continue;
		} else if (isdigit(last) || last == ',') {
		    ch = ',';
		}
	    } else if (ch == ',') {
		;
	    } else {
		if (last == ',')
		    result[k++] = (char) last;
		result[k++] = (char) ch;
	    }
	    last = ch;
	}
	result[k] = '\0';
    }
    return result;
}

static bool
comma_is_needed(const char *source)
{
    bool result = FALSE;

    if (source != 0) {
	unsigned len = strlen(source);
	if (len != 0)
	    result = (source[len - 1] != ',');
    } else {
	result = FALSE;
    }
    return result;
}

/*
 * Add a command-line parameter to the tab-list.  It can be blank- or comma-
 * separated (or a mixture).  For simplicity, empty tabs are ignored, e.g.,
 *	tabs 1,,6,11
 *	tabs 1,6,11
 * are treated the same.
 */
static const char *
add_to_tab_list(char **append, const char *value)
{
    char *result = *append;
    char *copied = trimmed_tab_list(value);

    if (copied != 0 && *copied != '\0') {
	const char *comma = ",";
	unsigned need = 1 + strlen(copied);

	if (*copied == ',')
	    comma = "";
	else if (!comma_is_needed(*append))
	    comma = "";

	need += strlen(comma);
	if (*append != 0)
	    need += strlen(*append);

	result = malloc(need);
	if (result != 0) {
	    *result = '\0';
	    if (*append != 0) {
		strcpy(result, *append);
		free(*append);
	    }
	    strcat(result, comma);
	    strcat(result, copied);
	}

	*append = result;
    }
    return result;
}

/*
 * Check for illegal characters in the tab-list.
 */
static bool
legal_tab_list(const char *program, const char *tab_list)
{
    bool result = TRUE;

    if (tab_list != 0 && *tab_list != '\0') {
	if (comma_is_needed(tab_list)) {
	    int n, ch;
	    for (n = 0; tab_list[n] != '\0'; ++n) {
		ch = UChar(tab_list[n]);
		if (!(isdigit(ch) || ch == ',' || ch == '+')) {
		    fprintf(stderr,
			    "%s: unexpected character found '%c'\n",
			    program, ch);
		    result = FALSE;
		    break;
		}
	    }
	} else {
	    fprintf(stderr, "%s: trailing comma found '%s'\n", program, tab_list);
	    result = FALSE;
	}
    } else {
	fprintf(stderr, "%s: no tab-list given\n", program);
	result = FALSE;
    }
    return result;
}

static void
usage(void)
{
    static const char *msg[] =
    {
	"Usage: tabs [options] [tabstop-list]"
	,""
	,"Options:"
	,"  -0       reset tabs"
	,"  -8       set tabs to standard interval"
	,"  -a       Assembler, IBM S/370, first format"
	,"  -a2      Assembler, IBM S/370, second format"
	,"  -c       COBOL, normal format"
	,"  -c2      COBOL compact format"
	,"  -c3      COBOL compact format extended"
	,"  -d       debug (show ruler with expected/actual tab positions)"
	,"  -f       FORTRAN"
	,"  -n       no-op (do not modify terminal settings)"
	,"  -p       PL/I"
	,"  -s       SNOBOL"
	,"  -u       UNIVAC 1100 Assembler"
	,"  -T name  use terminal type 'name'"
	,""
	,"A tabstop-list is an ordered list of column numbers, e.g., 1,11,21"
	,"or 1,+10,+10 which is the same."
    };
    unsigned n;

    fflush(stdout);
    for (n = 0; n < SIZEOF(msg); ++n) {
	fprintf(stderr, "%s\n", msg[n]);
    }
    ExitProgram(EXIT_FAILURE);
}

int
main(int argc, char *argv[])
{
    int rc = EXIT_FAILURE;
    bool debug = FALSE;
    bool no_op = FALSE;
    int n, ch;
    NCURSES_CONST char *term_name = 0;
    const char *mar_list = 0;	/* ignored */
    char *append = 0;
    const char *tab_list = 0;

    if ((term_name = getenv("TERM")) == 0)
	term_name = "ansi+tabs";

    /* cannot use getopt, since some options are two-character */
    for (n = 1; n < argc; ++n) {
	char *option = argv[n];
	switch (option[0]) {
	case '-':
	    while ((ch = *++option) != '\0') {
		switch (ch) {
		case 'a':
		    switch (*option) {
		    case '\0':
			tab_list = "1,10,16,36,72";
			/* Assembler, IBM S/370, first format */
			break;
		    case '2':
			tab_list = "1,10,16,40,72";
			/* Assembler, IBM S/370, second format */
			break;
		    default:
			usage();
		    }
		    break;
		case 'c':
		    switch (*option) {
		    case '\0':
			tab_list = "1,8,12,16,20,55";
			/* COBOL, normal format */
			break;
		    case '2':
			tab_list = "1,6,10,14,49";
			/* COBOL compact format */
			break;
		    case '3':
			tab_list = "1,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,67";
			/* COBOL compact format extended */
			break;
		    default:
			usage();
		    }
		    break;
		case 'd':	/* ncurses extension */
		    debug = TRUE;
		    break;
		case 'f':
		    tab_list = "1,7,11,15,19,23";
		    /* FORTRAN */
		    break;
		case 'n':	/* ncurses extension */
		    no_op = TRUE;
		    break;
		case 'p':
		    tab_list = "1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61";
		    /* PL/I */
		    break;
		case 's':
		    tab_list = "1,10,55";
		    /* SNOBOL */
		    break;
		case 'u':
		    tab_list = "1,12,20,44";
		    /* UNIVAC 1100 Assembler */
		    break;
		case 'T':
		    ++n;
		    if (*++option != '\0') {
			term_name = option;
		    } else {
			term_name = argv[n++];
		    }
		    option += ((int) strlen(option)) - 1;
		    continue;
		default:
		    if (isdigit(UChar(*option))) {
			tab_list = option;
			++n;
		    } else {
			usage();
		    }
		    option += ((int) strlen(option)) - 1;
		    break;
		}
	    }
	    break;
	case '+':
	    while ((ch = *++option) != '\0') {
		switch (ch) {
		case 'm':
		    mar_list = option;
		    break;
		default:
		    /* special case of relative stops separated by spaces? */
		    if (option == argv[n] + 1) {
			tab_list = add_to_tab_list(&append, argv[n]);
		    }
		    break;
		}
	    }
	    break;
	default:
	    if (append != 0) {
		if (tab_list != (const char *) append) {
		    /* one of the predefined options was used */
		    free(append);
		    append = 0;
		}
	    }
	    tab_list = add_to_tab_list(&append, option);
	    break;
	}
    }

    setupterm(term_name, STDOUT_FILENO, (int *) 0);

    max_cols = (columns > 0) ? columns : 80;

    if (!VALID_STRING(clear_all_tabs)) {
	fprintf(stderr,
		"%s: terminal type '%s' cannot reset tabs\n",
		argv[0], term_name);
    } else if (!VALID_STRING(set_tab)) {
	fprintf(stderr,
		"%s: terminal type '%s' cannot set tabs\n",
		argv[0], term_name);
    } else if (legal_tab_list(argv[0], tab_list)) {
	int *list = decode_tabs(tab_list);

	if (!no_op)
	    tputs(clear_all_tabs, 1, putch);

	if (list != 0) {
	    if (!no_op)
		do_tabs(list);
	    if (debug) {
		fflush(stderr);
		printf("tabs %s\n", tab_list);
		print_ruler(list);
		write_tabs(list);
	    }
	    free(list);
	} else if (debug) {
	    fflush(stderr);
	    printf("tabs %s\n", tab_list);
	}
	rc = EXIT_SUCCESS;
    }
    if (append != 0)
	free(append);
    ExitProgram(rc);
}
